Add database

ebogacheva opened the pull request.

ebogacheva committed:
feat: UserDbStorage and FriendshipDbStorage

ebogacheva committed:
ref: UserService

ebogacheva committed:
ref: UserService and Storage - exceptions

ebogacheva committed:
ref: UserController

ebogacheva committed:
feat: Services, Storages, Exceptions

ebogacheva committed:
ref: tests passed

ebogacheva committed:
ref: code style, services logic

ebogacheva committed:
fix: sql queries for friends requests and confirmed friends

ebogacheva committed:
ref: log constants
feat: confirmedFriends endpoint

ebogacheva committed:
ref: schema.sql -add constraints

ebogacheva committed:
feat: diagram, README.md

ebogacheva committed:
feat: UserDbStorageTest

ebogacheva committed:
feat: FilmDbStorageTest

ebogacheva committed:
feat: FriendshipDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

ebogacheva committed:
feat: FilmGenreDbStorageTest, FilmMpaDbStorageTest, LikeDbStorageTest

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @GetMapping("/{id}")
+    public Genre getGenreById(@PathVariable("id") int id) {
```
Тут можно валидировать id 
Пример - https://stackoverflow.com/questions/24582842/is-it-possible-to-validate-requestparam-in-spring-mvc-rest-endpoint

ebogacheva replied:
Сделала, спасибо.

db-exp commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/like/LikeDbStorageTest.java:
```diff
@@ -0,0 +1,115 @@
+class LikeDbStorageTest {
+
+    private static final String SQL_QUERY_DELETE_LIKES = "DROP TABLE likes";
+    private static final String SQL_QUERY_CREATE_LIKES =
```
можно вынести скрипт 
https://www.baeldung.com/spring-boot-data-sql-and-schema-sql

ebogacheva replied:
Сделала.

db-exp commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,80 @@
+  id int GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+  email varchar NOT NULL,
+  login varchar(255) NOT NULL,
+  name varchar(255),
```
для имени 255, кажется, что много

ebogacheva replied:
Пусть 50.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/utils/Validator.java:
```diff
@@ -48,7 +50,7 @@ public static void isValidated(User user) {
+        if (Objects.isNull(user.getName()) || user.getName().isBlank()) {
             user.setName(user.getLogin());
         }
         if (user.getBirthday().isAfter(LocalDate.now())) {
```
это можно решить аннотацией Past на поле

ebogacheva replied:
Это у меня не актуальный способ валидации. Сейчас все работает на аннотациях. Убрала лишний класс.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/genre/GenreDbStorage.java:
```diff
@@ -0,0 +1,36 @@
+    private final JdbcTemplate jdbcTemplate;
+
+    @Override
+    public Genre getGenreById(int genreId) {
```
лучше обрабатывать все ошибки здесь и возвращать наружу, например, Optional<Genre> - слои выше репозитория не должны знать об sql ошибках

db-exp replied:
А DataAccessException тут вообще лишний, потому что "это все остальные возможные исключения"
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/core/JdbcTemplate.html#queryForObject(java.lang.String,java.lang.Class)

ebogacheva replied:
Sql ошибки у всех getById отловила в storage'ах и отправила наверх пустые Optional на ошибки. 
В Service по пустым Optional отправила в контроллеры NOT_FOUND.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,34 @@
+        Genre genre;
+        try {
+            genre = genreDbStorage.getGenreById(id);
+        } catch (DataAccessException ex) {
```
оставил комментарий в storage

ebogacheva replied:
IncorrectResultSizeDataAccessException - теперь в Storage.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/model/Genre.java:
```diff
@@ -0,0 +1,19 @@
+    private String name;
+
+    @Override
+    public int compareTo(Genre genre) {
```
просто вопрос - а как используем? жанры же нельзя сравнить просто по id, если ужасы выше комедии только по id это не совсем верно, потому что вещь субъективная)

ebogacheva replied:
Тесты в Postman'e хотят жанры отсортированными.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -2,77 +2,110 @@
+        try {
+            film = filmDbStorage.getById(filmId);
+            log.info(Constants.GOT_FILM_BY_ID, filmId);
+        } catch (DataAccessException ex) {
```
см. комментарий в GenreStorage

ebogacheva replied:
Сделано.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -2,77 +2,110 @@
+            if (filmDbStorage.delete(filmId)) {
+                log.info(Constants.DELETED_FILM_LOG, filmId);
+            }
+        } catch (DataAccessException ex) {
```
см. комментарий в GenreStorage
А вообще - https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/core/JdbcTemplate.html#queryForObject(java.lang.String,java.lang.Class)
тут DataAccessException - как ошибка "на всякий случай" - предлагаю вообще от нее отказаться, но сам результат на уровне storage обернуть в Optional

ebogacheva replied:
Для delete и update отказалась от обработки исключений "на всякий случай". 
Но по результату delete и update, если соответствующие изменения не были внесены, отправляю из сервиса в контроллер NotPerformedFilmorateOperationException.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/service/MpaService.java:
```diff
@@ -0,0 +1,36 @@
+        Mpa mpa;
+        try {
+            mpa = mpaDbStorage.getMpaById(id);
+        } catch (DataAccessException ex) {
```
см. комментарий в GenreStorage

ebogacheva replied:
Сделано.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/service/MpaService.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final MpaStorage mpaDbStorage;
+
+    private final FilmMpaStorage filmMpaDbStorage;
```
Лучше ходить не в соседний storage, а в соседний сервис, то есть MpaService

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/filmMpa/FilmMpaDbStorage.java:
```diff
@@ -0,0 +1,51 @@
+        try {
+            mpa = jdbcTemplate.queryForObject(SQL_QUERY_GET_MPA_BY_FILM_ID,  FilmorateRowMappers::getMpa, filmId);
+            log.info(GOT_FILM_MPA_INFO, filmId);
+        } catch (DataAccessException ignored) {
```
см. комментарий в GenreStorage

ebogacheva replied:
Реорганизовала все storage-service.

Была проблема - получалась циклическая ссылка film-likes. 
Убрала LikeService - функционал переехал в FilmService. 
FilmService "работает" с двумя Storage - film и likes.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/filmGenre/FilmGenreDbStorage.java:
```diff
@@ -0,0 +1,49 @@
+            genres = jdbcTemplate.query(SQL_QUERY_GET_ALL_FILM_GENRES, FilmorateRowMappers::getGenre, filmId);
+            log.info(Constants.GOT_ALL_GENRES_FOR_FILM_LOG, filmId);
+            return genres;
+        } catch (DataAccessException ignored) {
```
см. комментарий в GenreStorage

ebogacheva replied:
Сделано.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public User getById(int id) {
+        return jdbcTemplate.queryForObject(SQL_QUERY_GET_USER_BY_ID, FilmorateRowMappers::getUser, id);
```
лучше обернуть в Optional

ebogacheva replied:
Сделано.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -17,78 +20,116 @@
+        List<User> users = new ArrayList<>();
+        try {
+            users = userDbStorage.findAll();
+        } catch (DataAccessException ignored) {
```
см. комментарий в GenreStorage, а в общем и целом совсем игнорировать исключения плохо
тут кмк оно совсем лишнее

ebogacheva replied:
Да, убрала, лишнее.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -17,78 +20,116 @@
+        try {
+            user = userDbStorage.getById(userId);
+            log.info(Constants.GOT_USER_BY_ID_LOG, userId);
+        } catch (DataAccessException ex) {
```
см. комментарий в GenreStorage

ebogacheva replied:
Исправила.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -17,78 +20,116 @@
+            if (userDbStorage.delete(id)) {
+                log.info(Constants.USER_DELETED_LOG, id);
+            }
+        } catch (DataAccessException ex) {
```
см. комментарий в GenreStorage

ebogacheva replied:
Исправила.

db-exp commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,125 @@
+        if (Objects.nonNull(film)) {
+            completeFilmWithMpaAndGenresFromDb(film);
+        }
+        return film;
```
лучше в Optional обернуть

ebogacheva replied:
Исправила.

ebogacheva committed:
fix: rearranged service and storage 'levels'
feat: add FriendshipService, delete LikeService (methods moved to FilmService)
fix: sql-exceptions are catched at storage level, optional goes further
fix: reduce number of custom exceptions

ebogacheva committed:
feat: sql-script for likesTest added

ebogacheva committed:
fix: sql-script for FilmDbStorageTest added
